types/core.go

```diff
@@ -0,0 +1,335 @@
+package types
+
+import (
+	"fmt"
+
+	"github.com/kong/deck/crud"
+	"github.com/kong/deck/konnect"
+	"github.com/kong/deck/state"
+	"github.com/kong/go-kong/kong"
+)
+
+func eventFromArg(arg crud.Arg) crud.Event {
+	event, ok := arg.(crud.Event)
+	if !ok {
+		panic("unexpected type, expected diff.Event")
+	}
+	return event
+}
+
+type Entity interface {
+	Type() string
+	CRUDActions() crud.Actions
+	PostProcessActions() crud.Actions
+}
+
+type entityImpl struct {
+	typ                string
+	cRUDActions        crud.Actions // needs to set client
+	postProcessActions crud.Actions // needs currentstate Set
+}
+
+func (e entityImpl) Type() string {
+	return e.typ
+}
+
+func (e entityImpl) CRUDActions() crud.Actions {
+	return e.cRUDActions
+}
+
+func (e entityImpl) PostProcessActions() crud.Actions {
+	return e.postProcessActions
+}
+
+type EntityOpts struct {
+	CurrentState  *state.KongState
+	TargetState   *state.KongState
+	KongClient    *kong.Client
+	KonnectClient *konnect.Client
+}
```

Comment:
This seems like it's named oddly: the state and client objects aren't _options_, correct? The comment indicating that the instances of this type will be de-duplicated in the future and the contents of the struct suggests that this is more connection config and states for the differ, not the individual entities it handles.

Reply:
`XxOpts` is a way to name a configuration struct to an abstraction, that's all. I can rename to `XxConfig` if that improves readability.

Reply:
Nah, if that's a standard naming pattern I should just get used to it then; disregard.